src/grpc/stream_infer_handler.cc

```diff
@@ -130,362 +130,424 @@ ModelStreamInferHandler::StartNewRequest()
-        state->context_->WriteResponseIfReady(state);
-      } else {
-        state->response_queue_->MarkNextResponseComplete();
-        state->complete_ = true;
-        state->context_->PutTaskBackToQueue(state);
-      }
-    }
+  // Read protobuf request, convert it into a Triton request, and execute it.
+  // Update any related state metadata, such as if model is decoupled or not.
+  // Send an error response if any error occurs.
+  PrepareAndSendTritonRequest(state);
 
-    // Now that the inference request is in flight, create a copy of
-    // 'state' and use it to attempt another read from the connection
-    // (i.e the next request in the stream).
-    State* next_read_state =
-        StateNew(tritonserver_.get(), context, Steps::READ);
+  // Now that the inference request is in flight, create a copy of
+  // 'state' and use it to attempt another read from the connection
+  // (i.e the next request in the stream).
+  State* next_read_state = StateNew(tritonserver_.get(), context, Steps::READ);
 
 #ifdef TRITON_ENABLE_TRACING
-    // Capture a timestamp for the time when we start waiting for this
-    // next request to read.
-    // Can't create trace as we don't know the model to be requested,
-    // track timestamps in 'state'
-    next_read_state->trace_timestamps_.emplace_back(std::make_pair(
-        "GRPC_WAITREAD_START", TraceManager::CaptureTimestamp()));
+  // Capture a timestamp for the time when we start waiting for this
+  // next request to read.
+  // Can't create trace as we don't know the model to be requested,
+  // track timestamps in 'state'
+  next_read_state->trace_timestamps_.emplace_back(
+      std::make_pair("GRPC_WAITREAD_START", TraceManager::CaptureTimestamp()));
 #endif  // TRITON_ENABLE_TRACING
 
-    next_read_state->context_->responder_->Read(
-        &next_read_state->request_, next_read_state);
+  next_read_state->context_->responder_->Read(
+      &next_read_state->request_, next_read_state);
 
-  } else if (state->step_ == Steps::COMPLETE) {
-    state->step_ = Steps::FINISH;
-    finished = true;
-  } else if (!state->is_decoupled_) {
-    // We handle the WRITTEN and WRITEREADY states little
-    // differently depending whether the inference request
-    // is for a decoupled model or not. This is because the
-    // grpc contract requires us to call Write() only once
-    // on a task. Hence, for decoupled writes, we call only
-    // one write and then wait for another notification from
-    // the completion queue to execute pending Write()'s, if
-    // any.
-
-    //
-    // Non-Decoupled state transitions
-    //
-    if (state->step_ == Steps::WRITTEN) {
-      state->context_->ongoing_write_ = false;
+  // Not finished with a request on the read step unless an error occurs above.
+  return false;
+}
+
+bool
+ModelStreamInferHandler::RequestCompleteStep(InferHandler::State* state)
+{
+  state->step_ = Steps::FINISH;
+  // Finished with the request.
+  return true;
```

Comment:
Also @tanmayv25 , I don't really understand the point of the COMPLETE state/step. Why don't we directly transition to FINISH state anywhere we would transition to COMPLETE instead? Seems like an unnecessary intermediate step.